### ISSUE TITLE
OPERATION_BASED_WITH_INTERFACES doesn't generate interface model as sealed interface

### ISSUE THREAD

**Issue by chalermpong** [apollographql/apollo-kotlin#4421]
### Summary
With `experimental_operationBasedWithInterfaces` codegen model, the interface model is not generated as `sealed interface` when it is inside another single model.

### Version
3.6.1-SNAPSHOT

### Steps to reproduce the behavior
Use this graphql
```graphql
# Schema
interface Animal {
  species: String!
  parent: Animal!
}

query TestOperation {
  animal {
    ... AnimalFragment
  }
}
fragment AnimalFragment on Animal {
  species
  parent {
    species
    ... on Lion {
      temperature
    }
    ... on Cat {
      name
    }
  }
}
```
The output code will be
```kotlin
public data class AnimalFragment(
  public val species: String,
  public val parent: Parent,
) : Fragment.Data {
  public interface Parent { // <-- Parent should be generated as sealed interface
    ...
  }
}
```

### Propose Fix
It is not generated as `sealed interface` because of this code:
```kotlin
// ModelBuilder
private val nestedBuilders = model.modelGroups.flatMap {
  it.models.map {
    ModelBuilder(
        context = context,
        model = it,
        superClassName = null,
        path = path + model.modelName,
        hasSubclassesInSamePackage = hasSubclassesInSamePackage, // <-- nestedBuilders use hasSubclassesInSamePackage from its parent model which is incorrect.
        adaptableWith = null
    )
  }
}
```
`hasSubclassesInSamePackage` should be calculated for each model instead of using parent's value.

**Comment by chalermpong** [comment]
@martinbonnin Right now, `ModelBuilder` apply `hasSubclassesInSamePackage` to nested model groups. I think `hasSubclassesInSamePackage` should be calculateed for each nested model group.
```kotlin
//ModelBuilder.kt
private val nestedBuilders = model.modelGroups.flatMap { modelGroup ->
    
    // Reuse localInheritance from FragmentModelsBuilder.kt
    val localInheritance = modelGroup.models.any { it.isInterface } &&
        modelGroup.models.any { !it.isInterface }

    modelGroup.models.map {
      ModelBuilder(
          context = context,
          model = it,
          superClassName = null,
          path = path + model.modelName,
          hasSubclassesInSamePackage = it.isInterface && localInheritance,
          adaptableWith = null
      )
    }
```
**Note**: This will also affect `responseBased` codegen output. I think it doesn't cause any problem. Please check this file [AnimalFragmentImpl.kt](https://github.com/chalermpong/apollo-kotlin/commit/162a2c3915b5d63691f43fd061d4a7d16cd56bbb#diff-03453714f654b31f4cbedf154de9d25ee49a84c476386303b0c0f6a13eb753e3)

Are you ok with this solution?

**Comment by chalermpong** [comment]
My previous solution wasn't work because of unwanted side effect in `responseBased`. 

Here is my final solution. Just remove `isInterface` from `hasSubclassesInSamePackage` calculation. Fragment in `responseBased` codegen always generated as interface, right?

```kotlin
// FragmentModelsBuilder
ModelBuilder(
    context = context,
    model = it,
    superClassName = if (addSuperInterface && it.id == fragment.dataModelGroup.baseModelId) KotlinSymbols.FragmentData else null,
    path = listOf(packageName),
    hasSubclassesInSamePackage = localInheritance,
    adaptableWith = null
)
```

**Comment by BoD** [comment]
Now closing as the new `operationBasedWithInterfaces` codegen is available in [3.7.0](https://github.com/apollographql/apollo-kotlin/releases/tag/v3.7.0)
